### PR TITLE
bulkio: Use checkpoint files to store backup job state.

### PR DESCRIPTION
**Pull request by miretskiy**
Use checkpoint file as a canonical location to store backup job
state and configuration.

Backup job no longer relies on system.jobs table to store
its state pertaing to the backup operation.  All information
necessary for backup job execution is persisted into
a (possibly encrypted) checkpoint file instead of the system.jobs
table.  

Release Notes: None

### CODE REVIEW

File: pkg/ccl/backupccl/backup_planning.go [hunk 1]
```diff
@@ -1192,16 +1199,9 @@ func backupPlanHook(
 		jr := jobs.Record{
 			Description: description,
 			Username:    p.User(),
-			DescriptorIDs: func() (sqlDescIDs []descpb.ID) {
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:1195]: I don't think we need to remove this in this change, so I'd leave it for now. We can decide later if we can stop setting this but there's no reason to couple it to the manifest-not-in-job change.

Comment (miretskiy) [reply pkg/ccl/backupccl/backup_planning.go:1195]: Done.

**Comment by ajwerner** [comment]
does this rely on consistency from the cloud storage implementation for correctness?

**Review by miretskiy — COMMENTED** [review]
@ajwerner Not in any way we don't already rely on it.   We already were using checkpoints before.  It's just we were creating checkpoints during resume phase instead of planning.

Checkpoint files are also checksummed (in addition to whatever cloud provider does), so we have that protection as well.

We don't rely on clouds providing any atomic operations, and there are some races possible (perhaps multiple backups into the same location) -- but this PR doesn't change those in any way.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57565#-:-MNk219xCLUBvCfqAAqz:bp2xj66)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

**Comment by dt** [comment]
@miretskiy well, in the past, we didn't really _rely_ on immediately seeing a prior write. Previously we would write the checkpoint but never really _needed_ to read it: we only tried to read it if the job was resumed, and even then, it was fine if we couldn't read it -- we'd just assume there was no prior progress and start backing up the spans from the job-stored manifest from scratch. Now we do rely on seeing our own write more or less immediately -- between planning the job and then it being adopted and started. We may need a delayed retry if we get a file does not exist error reading the checkpoint when we begin resume? On the bright side, other than that initial case, where we error if we don't see _any_ file, we're fine with not seeing the most recent write -- if we get a slightly older than most recent write we're still correct, we just re-do a bit of work. It is really only if we resume before we see the initial write that we have a problem.

**Comment by dt** [comment]
so to answer @ajwerner : yes and no -- we won't produce an _incorrect_ backup in the face of a, say, eventually consistent cloud store, but we may, as this PR currently stands, be unable to start the backup if we don't see our own immediately-prior creation of a file.

**Comment by miretskiy** [comment]
@dt I somewhat disagree. We are not reading and writing across regions. No
cloud provider could have an API that says "yup, I've created a file and
written data to it" just to come back later and say "nope, your file can't
be read in the same region because it's metadata is not there, but will be
there eventually".  This would be broken api and would break pretty much
every application including entirety of our database.

**Comment by dt** [comment]
I believe s3 used to have negative caching? IIRC if you requested a file that didn’t exist, then wrote it, a subsequent request would  sometimes continue to not see it for some short period? But they announced strong consistency last week and a quick survey of the other providers suggest they all have read-after-write consistency too so I’m not worried

**Review by pbardea — APPROVED** [review]
LGTM

Re: backport considerations, the only thing that comes to mind is that a backup job started on an old node and adopted one running this version would fail the backup since it wouldn't see the tmp checkpoint file. Although I don't think that running a backup during an upgrade is destined to succeed in any case (https://github.com/cockroachdb/cockroach/issues/57635).

> Reviewed 2 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57565#-:-MNxbKCe3gjjkRRAj7y7:b-oqfxhf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

**Review by dt — APPROVED** [review]

File: pkg/ccl/backupccl/backup_test.go [hunk 2]
```diff
@@ -1097,20 +1097,14 @@ func TestBackupRestoreSystemJobs(t *testing.T) {
 
 	const numAccounts = 0
 	_, _, sqlDB, _, cleanupFn := BackupRestoreTestSetup(t, MultiNode, numAccounts, InitNone)
-	conn := sqlDB.DB.(*gosql.DB)
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/backupccl/backup_test.go]: whoops, that was a leftover comment from previous review that I thought I'd removed -- carry on.

**Pull request merged**